authorization/v2/uniformResourceIdentifier.go

```diff
@@ -24,6 +24,64 @@ func NewUriFromRequest(resource resources.Resource, handler gin.HandlerFunc, ctx
 	}
 }
 
+// NewURIFromString parses the string representation of a URI into the UniformResourceIdentifier struct.
+// NewURIFromString expects the string to be of the following form, otherwise ErrInvalidURI is returned.
+// hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+func NewURIFromString(source string) (UniformResourceIdentifier, error) {
+	var (
+		arguments map[string]string
+		metadata  map[string]string
+	)
+
+	pathArgumentSplit := strings.Split(source, "?")
+	path := pathArgumentSplit[0]
+
+	if len(pathArgumentSplit) > 2 {
+		return UniformResourceIdentifier{}, ErrInvalidURI
+	}
+
+	if len(pathArgumentSplit) == 2 {
+		var err error
+
+		argumentsMetadataSplit := strings.Split(pathArgumentSplit[1], "#")
+		arguments, err = unmarshalArguments(argumentsMetadataSplit[0])
+		if err != nil {
+			return UniformResourceIdentifier{}, err
+		}
+
+		if len(argumentsMetadataSplit) > 1 {
+			metadata, err = unmarshalArguments(argumentsMetadataSplit[1])
+			if err != nil {
+				return UniformResourceIdentifier{}, err
+			}
+		}
+	}
+
+	return UniformResourceIdentifier{
+		path:      path,
+		arguments: arguments,
+		metadata:  metadata,
+	}, nil
+}
+
+// MarshalJSON will convert the UniformResourceIdentifier struct into the standard string representation for URIs.
+func (uri UniformResourceIdentifier) MarshalJSON() ([]byte, error) {
```

Comment:
If we're using `QueryUnescape` in `NewURIFromString` we should also use `QueryEscape` for the arguments and the metadata here.

https://golang.org/pkg/net/url/#QueryEscape